JBIDE-22304 Server Editor: Pod deployment path is not shown

scabanovich opened the pull request:
Pod path will be asynchronously loaded. When job is on, text field will have message 'loading...'. User can modify data/cancel/finish without waiting for loading to complete. If it is completed, the loaded path will be saved to server data. The loading is on only when 'Use inferred pod path' checkbox is selected. Checking it off stops the process and returns the previous pod path value.

fbricon commented:
Inferred pod paths should be cached while the Wizard is opened

scabanovich commented:
Ok, done.

adietish commented:
I think that the "Pod Deployment Path" widget should be disabled or at least non-writable as long as we're loading the path. I currently can write to it which seems weird.

adietish commented:
in the editor stuff now is weird:
The text widget with the pod path displays "Loading..." but is editable. Once the loading is finished, it replaces the "Loading.." with the actual value. If you now try to edit the value it gets regularly replaced by the inferred value. 
I think that we need a checkbox in the editor like we have it in the wizard. Being checked we are inferring the path and loading is happening while the text field is not editable. In the opposite case we're not inferring, the loading is not happening and the text widget is editable.

scabanovich commented:
@adietish , I added the checkbox to the editor. On editor start, it will be auto-enabled only if pod path is empty which means that user did not wait until it was loaded when creating server adapter.

fbricon commented:
I had an existing server, that I opened after the workspace started, without authenticating first. I got 
![](http://content.screencast.com/users/fbricon/folders/Jing/media/9bbd6858-0374-4d63-9c3a-fb98518f9dd4/00000426.png)

```
!MESSAGE An internal error occurred during: "Setting connection, deploy project...".
!STACK 0
com.openshift.restclient.authorization.ResourceForbiddenException: User "openshift-dev" cannot list services in project "php"
	at com.openshift.internal.restclient.DefaultClient.createOpenShiftException(DefaultClient.java:483)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:158)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:139)
	at org.jboss.tools.openshift.core.connection.Connection.getResources(Connection.java:451)
	at org.jboss.tools.openshift.core.server.OpenShiftServerUtils.getService(OpenShiftServerUtils.java:321)
	at org.jboss.tools.openshift.core.server.OpenShiftServerUtils.getService(OpenShiftServerUtils.java:308)
	at org.jboss.tools.openshift.internal.ui.server.OpenShiftServerEditorSection$13.run(OpenShiftServerEditorSection.java:579)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: com.openshift.internal.restclient.http.HttpClientException: {
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "User \"openshift-dev\" cannot list services in project \"php\"",
  "reason": "Forbidden",
  "details": {
    "kind": "services"
  },
  "code": 403
}

	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.createException(UrlConnectionHttpClient.java:278)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:211)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:187)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.get(UrlConnectionHttpClient.java:149)
	at com.openshift.internal.restclient.DefaultClient.list(DefaultClient.java:153)
	... 6 more
Caused by: java.io.IOException: Server returned HTTP response code: 403 for URL: https://10.1.2.2:8443/api/v1/namespaces/php/services
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1840)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at com.openshift.internal.restclient.http.UrlConnectionHttpClient.request(UrlConnectionHttpClient.java:207)
	... 9 more
```

Also it seems the backround for the checkbox label is grey-er than normal

scabanovich commented:
@fbricon , can exception be reproduced without this PR? Anyway, it is strange, because requesting resources from not authenticated connection should prompt connection wizard.

scabanovich commented:
The checkbox background in Ubuntu was as good as of other widgets.
Now I set the background taken form the parent composite, the view on Ubuntu remains good, I hope it will help on other OS.

fbricon commented:
So the exception on startup also occurs in master. I'm pretty sure this is a problem that @jcantrill fixed in the past. I'll open a separate bug.

fbricon commented:
Background still doesn't look right. I guess it's the same issue describe in https://eclipse.org/articles/Article-Forms/article.html, Figure 4.

fbricon commented:
So the exception was caused by me having deleted the openshift project beforehand.

Anyways, there are some other issues, more serious than the dark background : http://screencast.com/t/Di9L2vnhTSH. 
So : 
- project combo is pretty slow to load.
- the editor is marked as dirty when load is (semi) complete
- inferring the pod path never ends, see this exception in the background:

```
org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4533)
	at org.eclipse.swt.SWT.error(SWT.java:4448)
	at org.eclipse.swt.SWT.error(SWT.java:4419)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:799)
	at org.eclipse.swt.widgets.Widget.checkWidget(Widget.java:589)
	at org.eclipse.swt.widgets.Widget.getData(Widget.java:865)
	at org.eclipse.ui.internal.ide.handlers.PagePropertyTester.test(PagePropertyTester.java:36)
	at org.eclipse.core.internal.expressions.Property.test(Property.java:58)
	at org.eclipse.core.internal.expressions.TestExpression.evaluate(TestExpression.java:100)
	at org.eclipse.core.internal.expressions.CompositeExpression.evaluateAnd(CompositeExpression.java:53)
	at org.eclipse.core.internal.expressions.WithExpression.evaluate(WithExpression.java:75)
	at org.eclipse.core.internal.expressions.CompositeExpression.evaluateOr(CompositeExpression.java:68)
	at org.eclipse.core.internal.expressions.OrExpression.evaluate(OrExpression.java:22)
	at org.eclipse.core.internal.expressions.ReferenceExpression.evaluate(ReferenceExpression.java:71)
	at org.eclipse.ui.internal.services.EvaluationReference.evaluate(EvaluationReference.java:72)
	at org.eclipse.ui.internal.services.EvaluationReference.evaluate(EvaluationReference.java:97)
	at org.eclipse.ui.internal.services.EvaluationService.requestEvaluation(EvaluationService.java:294)
	at org.eclipse.ui.internal.handlers.DirtyStateTracker.update(DirtyStateTracker.java:45)
	at org.eclipse.ui.internal.handlers.DirtyStateTracker.propertyChanged(DirtyStateTracker.java:107)
	at org.eclipse.ui.part.WorkbenchPart.firePropertyChange(WorkbenchPart.java:120)
	at org.eclipse.wst.server.ui.internal.editor.ServerEditor.firePropertyChange(ServerEditor.java:436)
	at org.eclipse.wst.server.ui.internal.editor.ServerEditor$13.propertyChange(ServerEditor.java:738)
	at org.eclipse.wst.server.ui.internal.editor.GlobalCommandManager.firePropertyChangeEvent(GlobalCommandManager.java:134)
	at org.eclipse.wst.server.ui.internal.editor.GlobalCommandManager.setDirtyState(GlobalCommandManager.java:548)
	at org.eclipse.wst.server.ui.internal.editor.GlobalCommandManager.executeCommand(GlobalCommandManager.java:369)
	at org.eclipse.wst.server.ui.internal.editor.ServerResourceCommandManager.execute(ServerResourceCommandManager.java:54)
	at org.eclipse.wst.server.ui.editor.ServerEditorSection.execute(ServerEditorSection.java:73)
	at org.jboss.tools.openshift.internal.ui.server.OpenShiftServerEditorModel.setPodPath(OpenShiftServerEditorModel.java:271)
	at org.jboss.tools.openshift.internal.ui.server.OpenShiftServerEditorModel.setPodPath(OpenShiftServerEditorModel.java:263)
	at org.jboss.tools.openshift.internal.ui.server.InferPodPathJob.setLoadingMessage(InferPodPathJob.java:96)
	at org.jboss.tools.openshift.internal.ui.server.InferPodPathJob.run(InferPodPathJob.java:63)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
```

 I don't think we should load all the projects here. It should just be a readonly field of the openshift project. Similarly, I think should only load one service and disable its field. But that's for another JIRA. Right now we should focus on having the inferred path not crashing

fbricon commented:
Ok so now the checkbox has a white background, and the inferpodpathjob seems to never end. Either stuck on 
```
Thread [Worker-11] (Suspended)	
	owns: SocksSocketImpl  (id=325)	
	owns: HttpsClient  (id=326)	
	owns: DelegateHttpsURLConnection  (id=327)	
	owns: HttpsURLConnectionImpl  (id=328)	
	PlainSocketImpl.socketConnect(InetAddress, int, int) line: not available [native method]	
	SocksSocketImpl(AbstractPlainSocketImpl).doConnect(InetAddress, int, int) line: 350	
	SocksSocketImpl(AbstractPlainSocketImpl).connectToAddress(InetAddress, int, int) line: 206	
	SocksSocketImpl(AbstractPlainSocketImpl).connect(SocketAddress, int) line: 188	
	SocksSocketImpl.connect(SocketAddress, int) line: 392	
	SSLSocketImpl(Socket).connect(SocketAddress, int) line: 589	
	SSLSocketImpl.connect(SocketAddress, int) line: 668	
	SSLSocketImpl(BaseSSLSocketImpl).connect(SocketAddress) line: 173	
	HttpsClient(NetworkClient).doConnect(String, int) line: 180	
	HttpsClient(HttpClient).openServer(String, int) line: 432	
	HttpsClient(HttpClient).openServer() line: 527	
	HttpsClient.<init>(SSLSocketFactory, URL, Proxy, int) line: 264	
	HttpsClient.New(SSLSocketFactory, URL, HostnameVerifier, Proxy, boolean, int, HttpURLConnection) line: 367	
	DelegateHttpsURLConnection(AbstractDelegateHttpsURLConnection).getNewHttpClient(URL, Proxy, int) line: 191	
	DelegateHttpsURLConnection(HttpURLConnection).plainConnect0() line: 1105	
	DelegateHttpsURLConnection(HttpURLConnection).plainConnect() line: 999	
	DelegateHttpsURLConnection(AbstractDelegateHttpsURLConnection).connect() line: 177	
	DelegateHttpsURLConnection(HttpURLConnection).getInputStream0() line: 1513	
	DelegateHttpsURLConnection(HttpURLConnection).getInputStream() line: 1441	
	HttpsURLConnectionImpl.getInputStream() line: 254	
	UrlConnectionHttpClient.request(HttpMethod, URL, IMediaType, int, ParameterValueMap) line: 207	
	UrlConnectionHttpClient.request(HttpMethod, URL, IMediaType, int, Parameter...) line: 187	
	UrlConnectionHttpClient.get(URL, int) line: 149	
	DefaultClient.list(String, String, Map<String,String>) line: 153	
	DefaultClient.list(String, String) line: 139	
	Project.getResources(String) line: 77	
	PodDeploymentPathProvider.getPod(IService, IProject, Connection) line: 82	
	PodDeploymentPathProvider.load(IService, Connection) line: 56	
	InferPodPathJob.loadPodPath(IService) line: 147	
	InferPodPathJob.run(IProgressMonitor) line: 72	
	Worker.run() line: 55	
```

or waiting in Thread.sleep

fbricon commented on plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/InferPodPathJob.java:
```diff
@@ -0,0 +1,154 @@
+			return Status.OK_STATUS;
+		}
+		setLoadingMessage();
+		while(true) {
```
the job might never end

scabanovich replied:
There are 2 break conditions below. Cycle continues only while method loadPodPath() fails and stop() is not called.

scabanovich replied:
As user you can stop loading by checking off 'Use inferred' and then turn it on again. Does it work?

adietish commented:
a problem I found In the wizard: The displayed pod path isnt updated when you switch services

how to reproduce:
1) ASSERT: have at least 2 services in your OpenShift project
2)  EXEC: in explorer: launch server adapter wizard via context menu of your service
3) ASSERT: wizard shows up
4) EXEC: hit "Advanced >>" to see the inferred pod path
5) ASSERT: the inferred pod path for the selected service is displayed
6) EXEC: select the other service in services tree
7) ASSERT: inferred pod path text field shows "loading..." and then displays the correct pod path after some loading time
8) EXEC: switch back to the initial service 

Result:
The inferred pod path is not updated.
https://youtu.be/E9a75foBkpY

scabanovich commented:
@adietish , I cannot reproduce the problem with the steps. I rebased the commit and fixed some problems that appeared after logic of model a bit modified in recent commits.
Could you please debug InferPodPathJob.propertyChange(PropertyChangeEvent evt) when you do the step 8. The created inside thread should start and call setLoadedPath(path) which resets pod path in the model and that should be fired to ui. When I debug, all goes as expected.

adietish commented:
testPR

adietish commented:
@scabanovich I can still reproduce this issue. In the current code you have to wait for the "loading..." to get replaced with the actual value. Once this happens, switch to a different service in the "Services" tree. The pod path wont change.

adietish commented:
@scabanovich Unfortunately this still isnt always changing from a displayed path to "Loading..." and vice versa: https://youtu.be/3wmrOicT84w

scabanovich commented:
@adietish , when you return to a previously selected service, the value should be taken from cache and set to pod path without 'loading...'. Could you please debug InferPodPathJob.propertyChange(PropertyChangeEvent evt) setting breakpoint before you select a previously selected service? Asserts:
1. The method InferPodPathJob.propertyChange is called.
2. serviceUpdate is set to true;
3. newService is set to the just selected service.
4. The following 2 'if' conditions are met.
5. local viriable 'path' is set to the correct path.
6. Thread is started and in its run setLoadedPath(path) and then model.setInferredPodPath(path) is called.
7. UI is refreshed with the set value.
Please check, which of these asserts breaks first in your case.

robstryker commented:
This PR is way out of date
